draco/analysis/transform.py

```diff
@@ -1265,3 +1265,235 @@ def process_finish(self) -> Union[containers.SiderealStream, containers.RingMap]
+                except ValueError as e:
+                    raise ValueError(
+                        f"Got dataset {name} with shape {arr.shape} and weight "
+                        f"with shape {mask.shape}. If using a weight mask, all "
+                        "datasets must have the same shape/axes as the weights, "
+                        "or the weights must be able to be broadcasted."
+                    ) from e
+            else:
+                mask = np.zeros(arr.shape, dtype=bool)
+
+            if ds.distributed:
+                # Select the best axes to redistribute to
+                original_ax_id = ds.distributed_axis
+                sel_ax_id, red_ax_id = _select_axes(ds_axes, original_ax_id)
+                arr = arr.redistribute(sel_ax_id)
+                mask = mpiarray.MPIArray.wrap(mask, axis=original_ax_id).redistribute(
+                    sel_ax_id
+                )
+
+            # Apply downselections to this dataset and the mask
+            for ax, sel in self.selections.items():
+                try:
+                    ax_ind = ds_axes.index(ax)
+                except ValueError:
+                    continue
+                arr = _apply_sel(arr, sel, ax_ind)
+                mask = _apply_sel(mask, sel, ax_ind)
+
+            if ds.distributed:
+                # Distribute to the axis to be used for reduction
+                arr = arr.redistribute(red_ax_id)
+                mask = mask.redistribute(red_ax_id)
+
+            # Apply the optional mask and reduction op
+            # Masked arrays won't work with a MPIArray, so we have to use
+            # a ndarray view and cast back afterwards
+            if self.apply_mask:
+                arr = np.ma.array(arr.view(np.ndarray), mask=mask.view(np.ndarray))
+                reduced = op(arr, axis=apply_over, keepdims=True).data
+            else:
+                arr = arr.view(np.ndarray)
+                reduced = op(arr, axis=apply_over, keepdims=True)
+            # The weights provided are just a boolean mask which is zero wherever
+            # all values along the reduction axes are zero and one elsewhere
+            weight = np.all(~mask, axis=apply_over, keepdims=True)
+
+            if ds.distributed:
+                reduced = mpiarray.MPIArray.wrap(reduced, axis=red_ax_id)
+                # Redistribute back to the original axis
+                reduced = reduced.redistribute(original_ax_id)
+                weight = weight.redistribute(original_ax_id)
+
+            out[name][:] = reduced[:]
+
+            # Add the weights back in. If a weight mask was used, the resulting
+            # weights array will be the same for each dataset, so we don't have
+            # to care about overwriting it each time (other than performance)
+            if self.generate_mask:
+                out.weight[:] = weight[:]
+            else:
+                out.weight[:] = 1
```

Comment:
We just set the weights to unity if we're not using a mask because we don't care about the weights in the reduced dataset? Is that right?

Reply:
That's the idea, but I'm very open to thoughts about how best to handle the weights through a reduction like this, since it's a bit convoluted and maybe not super useful the way it's done here

Reply:
Hmm, it would depend on what we might want to do with the weights in a reduced dataset. You could consider allowing the user to also specify a reduction operation to be applied to the weights, in case we want to retain some of that information when examining the corresponding reduced dataset...but this could also wait until we've spent some time working with reduced products, so we know what would be useful

Reply:
I've done a fair bit of a rework with how the weights are handled, which I think should make more sense and be a bit more useful